mergify_engine/rules/__init__.py

```diff
@@ -35,30 +35,100 @@
-        )
-    except filter.InvalidQuery as e:
-        raise voluptuous.Invalid(
-            message=f"Invalid condition '{value}'. {str(e)}", error_message=str(e)
-        )
+@dataclasses.dataclass
+class RuleCondition:
+    condition: dataclasses.InitVar[typing.Union[str, filter.TreeT]]
+    description: typing.Optional[str] = None
+    partial_filter: filter.Filter = dataclasses.field(init=False)
+    match: bool = dataclasses.field(init=False, default=False)
+    evaluated: bool = dataclasses.field(init=False, default=False)
+    evaluation_error: typing.Optional[str] = dataclasses.field(init=False, default=None)
+
+    def __post_init__(self, condition: typing.Union[str, filter.TreeT]) -> None:
+        try:
+            if isinstance(condition, str):
+                self.partial_filter = filter.Filter.parse(condition)
+            else:
+                self.partial_filter = filter.Filter(condition)
+        except filter.parser.pyparsing.ParseException as e:
+            raise voluptuous.Invalid(
+                message=f"Invalid condition '{condition}'. {str(e)}",
+                error_message=str(e),
+            )
+        except filter.InvalidQuery as e:
+            raise voluptuous.Invalid(
+                message=f"Invalid condition '{condition}'. {str(e)}",
+                error_message=str(e),
+            )
 
+    def __str__(self):
+        # NOTE(sileht): Move _tree_to_str() here?
+        return str(self.partial_filter)
 
-RuleConditions = typing.NewType("RuleConditions", typing.List[filter.Filter])
-RuleMissingConditions = typing.NewType(
-    "RuleMissingConditions", typing.List[filter.Filter]
-)
+    def copy(self) -> "RuleCondition":
+        return RuleCondition(self.partial_filter.tree, self.description)
+
+    async def __call__(self, obj: filter.GetAttrObjectT) -> bool:
+        if self.evaluated:
+            raise RuntimeError("RuleCondition can be evaluated only once")
+        self.evaluated = True
+        try:
+            self.match = await self.partial_filter(obj)
+        except live_resolvers.LiveResolutionFailure as e:
+            self.match = False
+            self.evaluation_error = e.reason
+        return self.match
+
+    def get_attribute_name(self) -> str:
+        tree = self.partial_filter.tree.get("-", self.partial_filter.tree)
+        name = list(tree.values())[0][0]
+        if name.startswith(filter.Filter.LENGTH_OPERATOR):
+            return str(name[1:])
+        return str(name)
+
+
+@dataclasses.dataclass
+class RuleConditions:
+    conditions: typing.List[RuleCondition]
+    match: bool = dataclasses.field(init=False, default=False)
+    evaluated: bool = dataclasses.field(init=False, default=False)
+
+    async def eval(self, obj: filter.GetAttrObjectT) -> bool:
+        if self.evaluated:
+            raise RuntimeError("RuleCondition can be evaluated only once")
+        self.evaluated = True
+        self.match = await filter.Filter({"and": self.conditions})(obj)
```

Comment:
```suggestion
        self.match = await filter.Filter({"and": self.conditions})(obj)
        self.evaluated = True
```
Sounds better in case the `Filter` fails.

Reply:
I have added this boolean to ensure conditions are copied and never reused. It's not there to know if the Filter has worked or not.
It helped me to find all occurrences of `condition(...)` to replace them by `condition.copy()(...)`

Maybe we can find a better name.

Reply:
Fair enough!